### PR TITLE
fix(conversational_retrieval_chain.ts): keep the chat history context alongside the vector store context

### LINKED ISSUE #1800: Keep the chat history context
Hey everyone, I made a really cool change to our chain! I'll present the problem and then explain how I was able to fix it.

With the standard Langchain chain connected to a vector store (for context access) and with memory storage, we can have a super powerful chatbot. The problem is that, with the standard Langchain chain, the chat is only able to reason from the results of the vector store similarity search. This introduces us to the following problem:

If I send a message to the AI introducing my friend Bob and say he is 28 years old, the bot will reason from the input message, maybe consider some vector store context (if there is any) and give us a response.

Then, if I ask the bot how old Bob is, it will be unable to answer "28 years old," because the bot will try to search for Bob in the vector store, but Bob was only mentioned in the message history.

In the current memory implementation, we use only to reformulate the user's final questions into a clearer question. Nothing else. That's why, no matter how we introduce Bob to the AI in a message, it will be unable to give any information about Bob in the next one, even though we have the memory implementation.

To fix this, I overwrote the standard prompt template of one of the Langchain chain steps not only to rewrite the user's final question in a clearer way, but also to include any messages from the history that have some similarity to the question, if any.

Here are some tests I did to identify and explain the problem and also the tests I did after overwriting the standard prompt template of this Langchain chain step.

### UI
image

### Underlying prompt built by Langchain
image

As you can see, LangChain is recreating the question, turning it into a standalone one.

### Example: Without prior relevant chat history context
image

#### UI
image

#### Underlying prompt build by Langchain
image

## Final result
CleanShot 2023-06-28 at 19 08 23

### PR DESCRIPTION
**Pull request by joaopcm**
The `question_generator_template` in `conversational_retrieval_chain.ts` file was updated to provide more clarity and include an optional chat history excerpt. The template now instructs to return the conversation history excerpt that includes any relevant context to the question if it exists and rephrase the follow up question to be a standalone question. The updated template also includes the phrase "Standalone question and optional chat history excerpt:".

@jopcmelo

Fixes #1800

### CODE REVIEW

**Comment by joaopcm** [comment]
CC @nfcampos

**Comment by jacoblee93** [comment]
I like this, just worried about potential regressions for folks using it in production right now. Let me circulate it a bit with the Python crew - maybe adding it under a flag would be a good compromise.

Would also want to add some tests/comparisons/docs if possible but hold off for now.

File: langchain/src/chains/chat_vector_db_chain.ts [hunk 1]
```diff
@@ -11,12 +11,12 @@ import { CallbackManagerForChainRun } from "../callbacks/manager.js";
 // eslint-disable-next-line @typescript-eslint/no-explicit-any
 export type LoadValues = Record<string, any>;
 
-const question_generator_template = `Given the following conversation and a follow up question, rephrase the follow up question to be a standalone question.
```
Comment (jacoblee93) [inline langchain/src/chains/chat_vector_db_chain.ts:14]: This one is deprecated, no need to modify it

Comment (joaopcm) [reply langchain/src/chains/chat_vector_db_chain.ts:14]: Yes, that makes sense! Let me remove this change from this file

**Comment by jacoblee93** [comment]
Consensus is this is useful but definite worries about regression - let's go with a flag in `fromLLM`. How about `extractConversationContext` in `questionGeneratorChainOptions`?

**Comment by joaopcm** [comment]
> Consensus is this is useful but definite worries about regression - let's go with a flag in `fromLLM`. How about `extractConversationContext` in `questionGeneratorChainOptions`?

@jacoblee93, tt really is concerning to add, it might be something close to a breaking change. Let's go with a flag instead! I'll update the PR in a few hours :)

**Comment by jacoblee93** [comment]
> > Consensus is this is useful but definite worries about regression - let's go with a flag in `fromLLM`. How about `extractConversationContext` in `questionGeneratorChainOptions`?
> 
> @jacoblee93, tt really is concerning to add, it might be something close to a breaking change. Let's go with a flag instead! I'll update the PR in a few hours :)

Ok cool! Would be great to have a few integration tests too

File: langchain/src/chains/chat_vector_db_chain.ts (hunk 1 above)
Comment (joaopcm) [reply langchain/src/chains/chat_vector_db_chain.ts]: Done ;)

**Comment by joaopcm** [comment]
@jacoblee93, could you review it again, please?

I just added:
- new tests testing the behavior with and without `questionGeneratorChainOptions.extractConversationContext`
- updated the docs including this new feature

**Comment by jacoblee93** [comment]
Have tried it out a bit and I think there's a lot of edge cases to where I'm nervous about adding this new prompt directly into the framework - for example, the model hallucinated a bad standalone question here and gave an incorrect response:

```
  await chain.call({
    question:
      "I have a friend called Bob. He's 28 years old. He'd like to know what is the powerhouse of the cell?",
  });

  const res2 = await chain.call({
    question: "What is the airspeed velocity of an unladen swallow?",
  });
```

GPT-4 would probably help but given we can pass the prompt in as a `template` into the `questionGeneratorChainOptions` directly, to start I'd love to use this as an example of how folks can customize the chain in the docs. Does that sound alright to you?

**Comment by juzarantri** [comment]
> Have tried it out a bit and I think there's a lot of edge cases to where I'm nervous about adding this new prompt directly into the framework - for example, the model hallucinated a bad standalone question here and gave a confidently incorrect response:
> 
> ```
>   await chain.call({
>     question:
>       "I have a friend called Bob. He's 28 years old. He'd like to know what is the powerhouse of the cell?",
>   });
> 
>   const res2 = await chain.call({
>     question: "What is the airspeed velocity of an unladen swallow?",
>   });
>   // It returned "Mitochondria are the powerhouse of the cell"
> ```
> 
> GPT-4 would probably help, but it still breaks the abstraction a bit in that the generated question is also not technically a standalone question anymore.
> 
> Given we can pass the prompt in as a `template` into the `questionGeneratorChainOptions` directly, I think there's a more moderate starting point of adding this as an example of how folks can customize the chain in the docs.
> 
> Basically, we'd encourage people to try this and potentially tweak on their own:
> 
> ```
>   const chain = ConversationalRetrievalQAChain.fromLLM(
>     model,
>     vectorStore.asRetriever(),
>     {
>       memory: new BufferMemory({
>         memoryKey: "chat_history",
>         returnMessages: true,
>       }),
>       questionGeneratorChainOptions: {
>         template: `Given the following conversation and a follow up question, return the conversation history excerpt that includes any relevant context to the question if it exists and rephrase the follow up question to be a standalone question.
> 
> Chat History:
> {chat_history}
> Follow Up Input: {question}
> 
> Your answer should follow the following format:
> \`\`\`
> Use the following pieces of context to answer the users question.
> If you don't know the answer, just say that you don't know, don't try to make up an answer.
> ----------------
> <Relevant chat history excerpt as context here>
> 
> Standalone question: <Rephrased question here>
> \`\`\`
> 
> Your answer:`
>       },
>     }
>   );
> ```
> 
> Does that sound alright to you?


So how can we maintain the history of each user?
Suppose there are 10 user simultaneously using this and the chain is one so I think all the response would be based on all the chats till now right?

And I guess still this version isn't yet published

**Comment by its-anas** [comment]
> Have tried it out a bit and I think there's a lot of edge cases to where I'm nervous about adding this new prompt directly into the framework - for example, the model hallucinated a bad standalone question here and gave a confidently incorrect response:
> 
> 
> 
> ```
> 
>   await chain.call({
> 
>     question:
> 
>       "I have a friend called Bob. He's 28 years old. He'd like to know what is the powerhouse of the cell?",
> 
>   });
> 
> 
> 
>   const res2 = await chain.call({
> 
>     question: "What is the airspeed velocity of an unladen swallow?",
> 
>   });
> 
>   // It returned "Mitochondria are the powerhouse of the cell"
> 
> ```
> 
> 
> 
> GPT-4 would probably help, but it still breaks the abstraction a bit in that the generated question is also not technically a standalone question anymore.
> 
> 
> 
> Given we can pass the prompt in as a `template` into the `questionGeneratorChainOptions` directly, I think there's a more moderate starting point of adding this as an example of how folks can customize the chain in the docs.
> 
> 
> 
> Basically, we'd encourage people to try this and potentially tweak on their own:
> 
> ```
> 
>   const chain = ConversationalRetrievalQAChain.fromLLM(
> 
>     model,
> 
>     vectorStore.asRetriever(),
> 
>     {
> 
>       memory: new BufferMemory({
> 
>         memoryKey: "chat_history",
> 
>         returnMessages: true,
> 
>       }),
> 
>       questionGeneratorChainOptions: {
> 
>         template: `Given the following conversation and a follow up question, return the conversation history excerpt that includes any relevant context to the question if it exists and rephrase the follow up question to be a standalone question.
> 
> 
> 
> Chat History:
> 
> {chat_history}
> 
> Follow Up Input: {question}
> 
> 
> 
> Your answer should follow the following format:
> 
> \`\`\`
> 
> Use the following pieces of context to answer the users question.
> 
> If you don't know the answer, just say that you don't know, don't try to make up an answer.
> 
> ----------------
> 
> <Relevant chat history excerpt as context here>
> 
> 
> 
> Standalone question: <Rephrased question here>
> 
> \`\`\`
> 
> 
> 
> Your answer:`
> 
>       },
> 
>     }
> 
>   );
> 
> ```
> 
> 
> 
> Does that sound alright to you?

Will this really have a working memory? Which args to use with chain.call() then?

**Comment by joaopcm** [comment]
@juzarantri 
> So how can we maintain the history of each user?
Suppose there are 10 user simultaneously using this and the chain is one so I think all the response would be based on all the chats till now right?
>
> And I guess still this version isn't yet published

I'm not sure if I understand you. Why would you create only one chat for 10 or more users? In this case, Langchain isn't capable of keeping track of different human messages in its memory since it only stores messages from `human` and `ai`.

If you want to have many users using your chain, you'd have to create different session IDs for each of them so that each of them have their own chat history.

**Comment by joaopcm** [comment]
> Have tried it out a bit and I think there's a lot of edge cases to where I'm nervous about adding this new prompt directly into the framework - for example, the model hallucinated a bad standalone question here and gave a confidently incorrect response:
> 
> ```
>   await chain.call({
>     question:
>       "I have a friend called Bob. He's 28 years old. He'd like to know what is the powerhouse of the cell?",
>   });
> 
>   const res2 = await chain.call({
>     question: "What is the airspeed velocity of an unladen swallow?",
>   });
>   // It returned "Mitochondria are the powerhouse of the cell"
> ```
> 
> GPT-4 would probably help, but it still breaks the abstraction a bit in that the generated question is also not technically a standalone question anymore.
> 
> Given we can pass the prompt in as a `template` into the `questionGeneratorChainOptions` directly, I think there's a more moderate starting point of adding this as an example of how folks can customize the chain in the docs.
> 
> Basically, we'd encourage people to try this and potentially tweak on their own:
> 
> ```
>   const chain = ConversationalRetrievalQAChain.fromLLM(
>     model,
>     vectorStore.asRetriever(),
>     {
>       memory: new BufferMemory({
>         memoryKey: "chat_history",
>         returnMessages: true,
>       }),
>       questionGeneratorChainOptions: {
>         template: `Given the following conversation and a follow up question, return the conversation history excerpt that includes any relevant context to the question if it exists and rephrase the follow up question to be a standalone question.
> 
> Chat History:
> {chat_history}
> Follow Up Input: {question}
> 
> Your answer should follow the following format:
> \`\`\`
> Use the following pieces of context to answer the users question.
> If you don't know the answer, just say that you don't know, don't try to make up an answer.
> ----------------
> <Relevant chat history excerpt as context here>
> 
> Standalone question: <Rephrased question here>
> \`\`\`
> 
> Your answer:`
>       },
>     }
>   );
> ```
> 
> Does that sound alright to you?

@jacoblee93, that makes sense! This is how I ended up fixing this issue at @getproducthub. Let me update the PR to include only the examples then :)

**Comment by joaopcm** [comment]
> Will this really have a working memory? Which args to use with chain.call() then?

@its-anas, it does have a working memory. There's no secret with `chain.call()`. You just have to provide the question to the chain and it will consider relevant character history excerpt alongside the vector store context under the hood.

Let me know if I've answered your question :)

**Comment by its-anas** [comment]
> @its-anas, it does have a working memory. There's no secret with `chain.call()`. You just have to provide the question to the chain and it will consider relevant character history excerpt alongside the vector store context under the hood.
> 
> Let me know if I've answered your question :)


@joaopcm based on your explanation and @jacoblee93 code, then the following chain must have the memory working. But when i interact with it, it doesn't remember anything.

Shouldn't i pass chat_history somewhere? When i pass it to chain.call() i get an error. So when i should pass it?

```js
const question = 'Hi, my name is John';

const sanitizedQuestion = question.trim().replaceAll('\n', ' ');

const history = [
  ['Hi, my name is John', 'Hi John! How can I assist you today?'],
];

const chain = ConversationalRetrievalQAChain.fromLLM(
  new ChatOpenAI({
    temperature: 0,
    modelName: 'gpt-3.5-turbo',
  }),
  vectorStore.asRetriever(),
  {
    memory: new BufferMemory({
      memoryKey: 'chat_history',
      returnMessages: true,
    }),
    questionGeneratorChainOptions: {
      template: `Given the following conversation and a follow up question, return the conversation history excerpt that includes any relevant context to the question if it exists and rephrase the follow up question to be a standalone question.

                Chat History:
                {chat_history}
                Follow Up Input: {question}

                Your answer should follow the following format:
                \`\`\`
                Use the following pieces of context to answer the users question.
                If you don't know the answer, just say that you don't know, don't try to make up an answer.
                ----------------
                <Relevant chat history excerpt as context here>

                Standalone question: <Rephrased question here>
                \`\`\`

                Your answer:`,
    },
  },
);

const response = await chain.call({
  question: sanitizedQuestion,
  // chat_history: history, > this result in Error: input values have 2 keys, you must specify an input key or pass only 1 key as input
});

```

**Comment by joaopcm** [comment]
> > @its-anas, it does have a working memory. There's no secret with `chain.call()`. You just have to provide the question to the chain and it will consider relevant character history excerpt alongside the vector store context under the hood.
> > Let me know if I've answered your question :)
> 
> @joaopcm based on your explanation and @jacoblee93 code, then the following chain must have the memory working. But when i interact with it, it doesn't remember anything.
> 
> Shouldn't i pass chat_history somewhere? When i pass it to chain.call() i get an error. So when i should pass it?
> 
> ```js
> const question = 'Hi, my name is John';
> 
> const sanitizedQuestion = question.trim().replaceAll('\n', ' ');
> 
> const history = [
>   ['Hi, my name is John', 'Hi John! How can I assist you today?'],
> ];
> 
> const chain = ConversationalRetrievalQAChain.fromLLM(
>   new ChatOpenAI({
>     temperature: 0,
>     modelName: 'gpt-3.5-turbo',
>   }),
>   vectorStore.asRetriever(),
>   {
>     memory: new BufferMemory({
>       memoryKey: 'chat_history',
>       returnMessages: true,
>     }),
>     questionGeneratorChainOptions: {
>       template: `Given the following conversation and a follow up question, return the conversation history excerpt that includes any relevant context to the question if it exists and rephrase the follow up question to be a standalone question.
> 
>                 Chat History:
>                 {chat_history}
>                 Follow Up Input: {question}
> 
>                 Your answer should follow the following format:
>                 \`\`\`
>                 Use the following pieces of context to answer the users question.
>                 If you don't know the answer, just say that you don't know, don't try to make up an answer.
>                 ----------------
>                 <Relevant chat history excerpt as context here>
> 
>                 Standalone question: <Rephrased question here>
>                 \`\`\`
> 
>                 Your answer:`,
>     },
>   },
> );
> 
> const response = await chain.call({
>   question: sanitizedQuestion,
>   // chat_history: history, > this result in Error: input values have 2 keys, you must specify an input key or pass only 1 key as input
> });
> ```

@its-anas, try adding `inputKey` to `BufferMemory`:
```
memory: new BufferMemory({
    memoryKey: 'chat_history',
    inputKey: 'question',
    returnMessages: true,
  }),
```

**Comment by joaopcm** [comment]
@its-anas , also, you might want to provide the history messages directly to `BufferMemory` instead of passing it to the `chain.call`.

Something like this:
```ts
const pastMessages = [
  new HumanChatMessage("My name's Jonas"),
  new AIChatMessage("Nice to meet you, Jonas!"),
];

const memory = new BufferMemory({
  chatHistory: new ChatMessageHistory(pastMessages),
});
```

**Comment by its-anas** [comment]
@joaopcm Thanks for helping me about this. But  I tried to do it like you said and it's still not remembering the conversation.

Note: when i asked it about what inside the chat history, it answered with what is stored in the vector database instead of what inside chat history.


```js
import { AIChatMessage, HumanChatMessage } from 'langchain/schema';
import { BufferMemory, ChatMessageHistory } from 'langchain/memory';

const question = 'Question should be here';

const sanitizedQuestion = question.trim().replaceAll('\n', ' ');

const history = [
  ['Hi, my name is John', 'Hello John! How can I assist you today?'],
  [
    "I'm from New York, you?",
    'I am an AI assistant and do not have a physical location.',
  ],
  [
    "Ok, but you remember who i am and where i'm from?",
    "I'm sorry, but as an AI assistant, I don't have the ability to remember personal information about individuals.",
  ],
  [
    'But what inside chat_history?',
    "The chat history contains information about the website's pages, such as the contact page.",
  ],
];

const pastMessages = history.reduce(
  (acc, subArray) => [
    ...acc,
    new HumanChatMessage(subArray[0]),
    new AIChatMessage(subArray[1]),
  ],
  [],
);

const model = new ChatOpenAI({
  temperature: 0,
  modelName: 'gpt-3.5-turbo',
});

const chain = ConversationalRetrievalQAChain.fromLLM(
  model,
  vectorstore.asRetriever(),
  {
    memory: new BufferMemory({
      memoryKey: 'chat_history',
      inputKey: 'question',
      chatHistory: new ChatMessageHistory(pastMessages),
      returnMessages: true,
    }),
    questionGeneratorChainOptions: {
      template: `Given the following conversation and a follow up question, return the conversation history excerpt that includes any relevant context to the question if it exists and rephrase the follow up question to be a standalone question.

                Chat History:
                {chat_history}
                Follow Up Input: {question}

                Your answer should follow the following format:
                \`\`\`
                Use the following pieces of context to answer the users question.
                If you don't know the answer, just say that you don't know, don't try to make up an answer.
                ----------------
                <Relevant chat history excerpt as context here>

                Standalone question: <Rephrased question here>
                \`\`\`

                Your answer:`,
    },
  },
);

const response = await chain.call({
  question: sanitizedQuestion,
});

```

**Comment by jacoblee93** [comment]
You'd need to initialize `ChatMessageHistory` like this: 
```
import { BufferMemory, ChatMessageHistory } from "langchain/memory";
import { HumanChatMessage, AIChatMessage } from "langchain/schema";

const pastMessages = [
  new HumanChatMessage("My name's Jonas"),
  new AIChatMessage("Nice to meet you, Jonas!"),
];

const memory = new BufferMemory({
  chatHistory: new ChatMessageHistory(pastMessages),
});
```

Let's move help requests to Discord though.

**Comment by jacoblee93** [comment]
@joaopcm yeah let's just add it to the docs as an example for now - I think this issue might require more structural changes to fix in a robust way.

**Comment by its-anas** [comment]
Thanks @joaopcm, i updated Langchain and memory is working. Thanks @jacoblee93 for you answer. Sorry guys for ruining your PR comments.

**Comment by joaopcm** [comment]
Sounds good, dude! Thanks!

**Comment by jacoblee93** [comment]
> Sounds good, dude! Thanks!

Thank you for pushing hard on this!

**Comment by SantoshKumarRavi** [comment]
yes. solution working fine.
but while streaming chat-history and rephrased question also in the token.
how to stream the output only as token ?

  ```
const model = new ChatOpenAI({
      openAIApiKey: this.config.get('openai.keys'),
      temperature: 0,
      modelName: modelName,
      streaming: true,
      callbacks: [
        {
          handleLLMNewToken(token: string) {
            console.log({ token });
          },
        },
      ],
    });
```

my issue is more or like this 
https://github.com/langchain-ai/langchain/issues/7599

**Pull request merged**